### PR TITLE
Stability

### PR DESCRIPTION
**Pull request by sumny**
see #31 
Potentially fixes points 2, 3, 4.

### CODE REVIEW

**Comment by sumny** [comment]
Summarizing this PR:
* `AcqOptimizer` gain a `ParamSet` containing `fix_distance` (`ParamLgl`) and `dist_threshold` (`ParamDbl`).
If `fix_distance` is `TRUE` (default is `FALSE`) the following will be done during `$optimize`:
Proposed points (`xdt`) are replaced by randomly sampled ones (or maybe augmented lhs later) if their Gower distance (calculated using the `gower` package) with respect to other proposed points or previously evaluated points falls below the `dist_threshold` parameter (see `fix_xdt_distance` helper function).
Additionally, `self$optimizer$optimize` is wrapped in a `tryCatch` statement where errors are logged and then returned with the error class upgraded to `"leads_to_exploration_error"` (this one is relevant later again), `"optimize_error"`, `"error"`, `"condition"`.

* `Surrogate` gains a `ParamSet` (that can be overloaded by subclasses, relevant for `SurrogateSingleCritLearner` and `SurrogateMultiCritLearners`) and a private `.insample_perf` field. Here the insample performance of the `Learner` will later be stored (if calculated). There is also an active binding for asserting the `.insample_perf` that can be overloaded (relevant later). Importantly, `$update` now proceeds to call a private `.update` method that should be implemented by subclasses. This way we can wrap the `.update` method in a `tryCatch` statement where errors are logged and then returned with the error class upgraded to `"update_error"` (this one is relevant later again), `"optimize_error"`, `"error"`, `"condition"`.
* `SurrogateSingleCritLearner` holds `calc_insample_perf` (`ParamLgl`), `perf_measure` (`ParamUty`) and `perf_threshold` (`ParamDbl`) as parameters. If `calc_insample_perf` is `TRUE`  (default is `FALSE`), during `.update` the insample performance (calculated via the `perf_measure` `Measure`)  of the `Learner` will be calculated and stored in the `.insample_perf` field. Also, it will be asserted whether the performance matches the `perf_threshold` (if not, an error is raised during `.update` and caught by the `$update` method in the superclass).
* `SurrogateMultiCritLearners` holds `calc_insample_perf` (`ParamLgl`), `perf_measures` (`ParamUty`) and `perf_thresholds` (`ParamUty`) as parameters. Functionality is the same as for `SurrogateSingleCritLearner` i.e., if `calc_insample_perf` is `TRUE` (default is `FALSE`) the insample performance of all Learners will be stored and all will be asserted whether they match their respective threshold.

All in all this allows for the following stability enhancements in loops, e.g. (`bayesopt_soo`):
In the `repeat` loop we will typically do:
```
acq_function$surrogate$update(xydt = archive_xy(archive), y_cols = archive$cols_y)
acq_function$update(archive) 
xdt = acq_optimizer$optimize(acq_function)
instance$eval_batch(xdt)
```

We can now wrap this in a `tryCatch` statement catching `leads_to_exploration_error` errors (triggered if the optimize step of the `AcqOptimizer` or the update (and potential insample performance assertion) step of the `Surrogate` fails and replacing `xdt` by a randomly sampled point (or maybe augmented lhs later):
```
xdt = tryCatch({
  acq_function$surrogate$update(xydt = archive_xy(archive), y_cols = archive$cols_y)
  acq_function$update(archive)
  acq_optimizer$optimize(acq_function, archive = archive)
}, leads_to_exploration_error = function(leads_to_exploration_error_condition) {
    lg$info("Proposing a randomly sampled point") 
    SamplerUnif$new(instance$search_space)$sample(1L)$data 
})
instance$eval_batch(xdt)
```

These additional 5 lines of codes can be easily integrated in any other loop.

By default, the Gower distance stuff of the `AcqOptimizer` and the insample performance assertion of the `Surrogate` is turned off. In this case this is more or less the normal routine except for if the update step of the `Surrogate` or the optimize step of the `AcqOptimizer` fails, we propose a randomly sampled point.

File: R/Surrogate.R [hunk 1]
```diff
@@ -27,9 +27,16 @@ Surrogate = R6Class("Surrogate",
     #' @param y_cols (`character()`)\cr
     #' Names of response columns.
     #'
-    #' @return `NULL`
+    #' @return `NULL`.
     update = function(xydt, y_cols) {
-      stop("Abstract")
+      tryCatch(private$.update(xydt, y_cols = y_cols),
```
Comment (jakob-r) [inline R/Surrogate.R:32]: assuming we use an mlr3learner I could also encapuslate there and use a fallback learner? is that more elegant?

File: R/AcqOptimizer.R [hunk 2]
```diff
@@ -15,23 +15,80 @@ AcqOptimizer = R6Class("AcqOptimizer",
...
         instance = OptimInstanceSingleCrit$new(objective = acq_function, terminator = self$terminator)
       } else {
         if (!"multi-crit" %in% self$optimizer$properties) {
           stopf("Optimizer %s is not multi-crit compatible but %s is multi-crit.", self$self$optimizer$format(), acq_function$id)
         }
         instance = OptimInstanceMultiCrit$new(objective = acq_function, terminator = self$terminator)
       }
-      self$optimizer$optimize(instance)
+
+      xdt = tryCatch(self$optimizer$optimize(instance),
```
Comment (jakob-r) [inline R/AcqOptimizer.R:55]: Here we don't know if the optimizer crashed or the surrogates' prediction?

Comment (sumny) [reply R/AcqOptimizer.R:63]: I agree, however, the error message is logged and should be different for surrogate vs. acq optimization failure

**Comment by sumny** [comment]
closes #31

**Pull request merged**